common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [IntervalIterator], [VoidsIterator], [DifferenceIterator] implementations.
+
+use crate::{
+    interval::{
+        IncorrectOrEmptyRangeError, IncorrectRangeError, NewWithLenError, OutOfBoundsError,
+    },
+    Interval, Numerated,
+};
+use core::{
+    fmt::{self, Debug, Display, Formatter},
+    ops::{Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive},
+};
+use num_traits::bounds::{LowerBounded, UpperBounded};
+
+/// Describes interval `start..=end`, which can be empty.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, derive_more::From)]
+pub struct IntervalIterator<T>(Option<Interval<T>>);
+
+impl<T: Numerated> IntervalIterator<T> {
+    /// New empty interval.
+    pub fn new_empty() -> Self {
+        Self(None)
+    }
+    /// Returns whether interval is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_none()
+    }
+    /// Returns inner value.
+    pub fn inner(&self) -> Option<Interval<T>> {
+        self.0
+    }
+}
+
+impl<T: Numerated> From<Interval<T>> for IntervalIterator<T> {
+    fn from(interval: Interval<T>) -> Self {
+        Self(Some(interval))
+    }
+}
+
+impl<T: Numerated> From<T> for IntervalIterator<T> {
+    fn from(point: T) -> Self {
+        Interval::from(point).into()
+    }
+}
+
+impl<T: Numerated + LowerBounded> From<RangeToInclusive<T>> for IntervalIterator<T> {
+    fn from(range: RangeToInclusive<T>) -> Self {
+        Interval::from(range).into()
+    }
+}
+
+impl<T: Numerated + UpperBounded, I: Into<T::Bound>> From<RangeFrom<I>> for IntervalIterator<T> {
+    fn from(range: RangeFrom<I>) -> Self {
+        Interval::try_from(range)
+            .map(Into::into)
+            .unwrap_or(Self(None))
+    }
+}
+
+impl<T: Numerated + UpperBounded + LowerBounded> From<RangeFull> for IntervalIterator<T> {
+    fn from(range: RangeFull) -> Self {
+        Interval::from(range).into()
+    }
+}
+
+impl<T: Numerated + LowerBounded + UpperBounded, I: Into<T::Bound>> From<RangeTo<I>>
+    for IntervalIterator<T>
+{
+    fn from(range: RangeTo<I>) -> Self {
+        Interval::try_from(range)
+            .map(Into::into)
+            .unwrap_or(Self(None))
+    }
+}
+
+impl<T, S, E> TryFrom<(S, E)> for IntervalIterator<T>
+where
+    T: Numerated + UpperBounded,
+    S: Into<T::Bound>,
+    E: Into<T::Bound>,
+{
+    type Error = IncorrectRangeError;
+
+    fn try_from(range: (S, E)) -> Result<Self, Self::Error> {
+        match Interval::try_from(range) {
+            Ok(interval) => Ok(interval.into()),
+            Err(IncorrectOrEmptyRangeError::EmptyRange(_)) => Ok(Self(None)),
```

Comment:
add cast of range errors to interval iterator for all of these constructors

Reply:
We need to make 3 impl for EmptyRangeError, for IncorrectOrEmptyRangeError, for NewWithLenError, and use them in 4 places. Are you sure we have to do this ?